README.md
# K8sProbe

Provides configurable HTTP liveness and readiness endpoints, intended to be used to support
[Kubernetes probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).

`K8sProbe` ships with a built-in `Cowboy` HTTP server making this module easy to integrate into
any project.

## Usage

### Installation

First, add `:k8s_probe` to your `mix.exs` dependencies like so:

```elixir
defp deps do
  [
    # … your other various dependencies
    {:k8s_probe, "0.3.0"}
  ]
end
```

If necessary run `mix deps.get` and `mix deps.compile` at this point to fetch and compile
`K8sProbe`.

Next, add `K8sProbe` to your supervision tree (perhaps in your `application.ex` file).  It would
look something like this:

```elixir
defmodule MyApp.Application do
  use Application

  def start(_type, _args) do
    children = [
      K8sProbe
    ]

    opts = [strategy: :one_for_one, name: MyApp.Supervisor]
    Supervisor.start_link(children, opts)
  end
end
```

Now you can try it out by launching your application and making a request to the probe endpoints
as follows:

```sh
# Start your app in iex
$ iex -S mix

# In another terminal
$ curl localhost:9991/readiness
OK
$ curl localhost:9991/liveness
OK
```

Congratulations, you have a default probe up and running!

### Customising the probe

This module ships with a completely minimal default probe that does nothing other than respond
with a `200 OK`, regardless of the state of your application.

This might be fine for simple applications, but for a more detailed implementation you need to
provide your own probe implementing the `K8sProbe.Probe` behaviour.  Here's an example:

```elixir
defmodule MyApp.MyProbe do
  @behaviour K8sProbe.Probe
  def readiness, do: :ok
  def liveness, do: :ok
end
```

Each of the probe functions must return either `:ok` or `:error`.  How you implement your probe
really depends on your application.

Having written your probe, you can configure K8sProbe to use it by passing it in the configuration.
Here's what your `application.ex` file might look like:

```elixir
defmodule MyApp.Application do
  use Application

  def start(_type, _args) do
    children = [
      {K8sProbe, probe_module: MyApp.MyProbe}
    ]

    opts = [strategy: :one_for_one, name: MyApp.Supervisor]
    Supervisor.start_link(children, opts)
  end
end
```

## Configuration

Generally it's recommended to just use the default configuration; easier isn't it!  But if you
must then read on...

### Port

By default K8sProbe listens on port 9991.  You may override it by passing it as config option.

## Configuring probes in Kubernetes

Here's an example Kubernetes deployment in YAML format that calls the probes as implemented by
default:

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-service
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/name: my-service
  template:
    spec:
      containers:
        name: my-service
        image: my-service
        ports:
          - containerPort: 9991
            name: liveness-port
            protocol: TCP
        readinessProbe:
          httpGet:
            path: /readiness
            port: liveness-port
        livenessProbe:
          httpGet:
            path: /liveness
            port: liveness-port
        startupProbe:
          httpGet:
            path: /readiness
            port: liveness-port
```

You'll need to modify this configuration before it will work of course, replacing the image with
your own.

Configuring Kubernetes to make use of the probes provided by this module is a topic unto itself.
See
[here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)
for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 0.3.1

### Changed

* Minor simplification of code.

## 0.2.0

### Added

* Better documentation

### Removed

* Dedicated health route is removed - use `/liveness` instead

## 0.1.0

* Initial release


test/endpoint_test.exs
defmodule K8sProbe.EndpointTest do
  use ExUnit.Case, async: true
  use Plug.Test

  defmodule GoodProbe do
    @behaviour K8sProbe.Probe
    def liveness, do: :ok
    def readiness, do: :ok
  end

  defmodule BadProbe do
    @behaviour K8sProbe.Probe
    def liveness, do: :error
    def readiness, do: :error
  end

  test "responds to liveness with 200 on success" do
    conn = conn(:get, "/liveness")

    conn = K8sProbe.Endpoint.call(conn, probe_module: GoodProbe)

    assert conn.state == :sent
    assert conn.status == 200
    assert conn.resp_body == "OK"
  end

  test "responds to liveness with 503 on error" do
    conn = conn(:get, "/liveness")

    conn = K8sProbe.Endpoint.call(conn, probe_module: BadProbe)

    assert conn.state == :sent
    assert conn.status == 503
    assert conn.resp_body == "Service Unavailable"
  end

  test "responds to readiness with 200 on success" do
    conn = conn(:get, "/readiness")

    conn = K8sProbe.Endpoint.call(conn, probe_module: GoodProbe)

    assert conn.state == :sent
    assert conn.status == 200
    assert conn.resp_body == "OK"
  end

  test "responds to readiness with 503 on error" do
    conn = conn(:get, "/readiness")

    conn = K8sProbe.Endpoint.call(conn, probe_module: BadProbe)

    assert conn.state == :sent
    assert conn.status == 503
    assert conn.resp_body == "Service Unavailable"
  end

  test "responds to other routes with 404" do
    conn = conn(:get, "/non-existant")

    conn = K8sProbe.Endpoint.call(conn, [])

    assert conn.state == :sent
    assert conn.status == 404
    assert conn.resp_body
  end
end


mix.exs
defmodule K8sProbe.MixProject do
  use Mix.Project

  @version "0.3.1"
  @repo_url "https://github.com/joeapearson/elixir-k8s-probe"

  def project do
    [
      app: :k8s_probe,
      version: @version,
      elixir: "~> 1.10",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      package: package(),
      docs: docs()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:plug_cowboy]
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false},
      {:nimble_options, "~> 0.3.0"},
      {:plug_cowboy, "~> 2.3.0"}
      # {:dep_from_hexpm, "~> 0.3.0"},
      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"}
    ]
  end

  defp package do
    [
      description:
        "Provides configurable HTTP liveness and readiness endpoints for supporting Kubernetes probes.",
      licenses: ["MIT"],
      links: %{"GitHub" => @repo_url}
    ]
  end

  defp docs do
    [
      main: "readme",
      extras: [
        "README.md",
        "CHANGELOG.md",
        "LICENSE"
      ],
      authors: [
        "Joe Pearson"
      ],
      source_ref: "v#{@version}",
      source_url: @repo_url,
      api_reference: false
    ]
  end
end


lib/k8s_probe.ex
defmodule K8sProbe do
  use Supervisor

  @config_schema [
    port: [
      type: :pos_integer,
      default: 9991
    ],
    probe_module: [
      type: :atom,
      default: K8sProbe.Probe
    ]
  ]

  @doc """
  Starts an instance of K8sProbe.

  ## Options

  #{NimbleOptions.docs(@config_schema)}
  """
  def start_link(opts \\ []) do
    validated_opts = NimbleOptions.validate!(opts, @config_schema)
    Supervisor.start_link(__MODULE__, validated_opts, name: __MODULE__)
  end

  @impl true
  def init(opts) do
    probe_module = Keyword.get(opts, :probe_module)
    port = Keyword.get(opts, :port)

    children = [
      Plug.Cowboy.child_spec(
        scheme: :http,
        plug: {K8sProbe.Endpoint, [probe_module: probe_module]},
        options: [port: port]
      )
    ]

    Supervisor.init(children, strategy: :one_for_one)
  end
end


lib/k8s_probe/endpoint.ex
defmodule K8sProbe.Endpoint do
  @moduledoc """
  A plug responsible for handling kubernetes liveness and readiness check probe requests.
  """

  use Plug.Router

  # :match plug matches routes and forwards them to :dispatch plug below
  plug(:match)
  # :dispatch plug attaches builder_opts() which allows us to configure the plug
  plug(:dispatch, builder_opts())

  # Liveness endpoint
  get "/liveness" do
    probe = get_probe_module!(opts)
    check_probe(conn, &probe.liveness/0)
  end

  # Readiness endpoint
  get "/readiness" do
    probe = get_probe_module!(opts)
    check_probe(conn, &probe.readiness/0)
  end

  # Responds 404 Not Found to any other path
  match _ do
    send_resp(conn, 404, "Not Found")
  end

  # Calls the specified probe
  defp check_probe(conn, fun) do
    send_response(conn, fun.())
  end

  # Dispatches a 200 OK response since the probe has responded :ok
  defp send_response(conn, :ok) do
    send_resp(conn, Plug.Conn.Status.code(:ok), "OK")
  end

  # Dispatches 503 Service Unavailable since the probe has responded with something other than :ok
  defp send_response(conn, _) do
    send_resp(conn, 503, "Service Unavailable")
  end

  # Gets the probe module from the options provided to this plug
  defp get_probe_module!(probe_module: probe_module) do
    probe_module
  end
end
